gettsim/interface.py

```diff
@@ -65,13 +69,87 @@ def compute_taxes_and_transfers(
         DataFrame containing computed variables.
 
     """
+
+    # Set defaults for some parameters
     targets = DEFAULT_TARGETS if targets is None else targets
     targets = parse_to_list_of_strings(targets, "targets")
     columns_overriding_functions = parse_to_list_of_strings(
         columns_overriding_functions, "columns_overriding_functions"
     )
     params = {} if params is None else params
 
+    # Create the dag
+    dag = set_up_dag(
+        data,
+        params,
+        functions,
+        targets,
+        columns_overriding_functions,
+        check_minimal_specification,
+        rounding,
+    )
+
+    # We delay the data preparation as long as possible such that other checks can fail
+    # before this
+    data = data.copy(deep=True)
+    data = _process_data(data)
+    data = _reduce_data(data)
+    ids = _dict_subset(data, set(data) & {"hh_id", "tu_id"})
+
+    results = execute_dag(dag, data, targets, debug)
+
+    results = _expand_data(results, ids)
+    results = pd.DataFrame(results)
+
+    if not debug:
+        results = results[targets]
+
+    results = _reorder_columns(results)
+
+    return results
+
+
+def set_up_dag(
+    data,
```

Comment:
But almost trivially, no? Go through all nodes of the DAG, check if it is a function or a data column, do the check if the latter. 

Given that the function is new, I would go that far in case I am not missing anything.

Reply:
Usual data columns are trivial. The problem is data columns that ovewrite functions. For those, the expected data type, depends on the return type of the function, but the function is not part of the DAG and doesn't exist outside of the function as it is implemented right now.

Reply:
Okay, just so that I understand: we would need to update the dict of expected data types defined for the input columns (can't look up right now) while looking at columns overriding functions?

What about passing (example, untested)

`{col: dtype for col, dtype in data.dtypes}`

for the moment?

I'd just like to signal that it has nothing to do with the data, really.

Reply:
I now restructured `compute_taxes_and_transfers` which now calls (among other) two functions:

- ``check_data_check_functions_and_merge_functions``
- ``prepare_functions_and_set_up_dag``

``prepare_functions_and_set_up_dag`` no longer requires ``data`` as argument. ``check_data_check_functions_and_merge_functions`` still does so (among others because `check_if_series_has_internal_type` requires the pd.series and not just the dtypes). 

But this structure should make clear that creating the DAG has nothing to do with data as was your concern if I understand correctly.

Reply:
Cool. Just a small follow-up re the expected data left now.